pkg/controllers/certificate/approver/agent_csr_approving.go

```diff
@@ -0,0 +1,289 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/karmada-io/karmada/pkg/util/certificate"
+)
+
+const (
+	csrApprovingController = "agent-csr-approving"
+	agentCSRGroup          = "system:karmada:agents"
+	agentCSRUserPrefix     = "system:karmada:agent:"
+)
+
+// AgentCSRApprovingController is used to automatically approve the agent's CSR.
+type AgentCSRApprovingController struct {
+	client      kubernetes.Interface
+	recognizers []csrRecognizer
+}
+
+// NewAgentCSRApprovingController return a NewAgentCSRApprovingController.
+func NewAgentCSRApprovingController(client kubernetes.Interface) *AgentCSRApprovingController {
+	return &AgentCSRApprovingController{
+		client:      client,
+		recognizers: recognizers(),
+	}
+}
+
+// SetupWithManager creates a controller and registers to controller manager.
+func (a *AgentCSRApprovingController) SetupWithManager(mgr controllerruntime.Manager) error {
+	var predicateFunc = predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			csr := e.Object.(*certificatesv1.CertificateSigningRequest)
+			// agent certificate is signed by "kubernetes.io/kube-apiserver-client" signer
+			return isIssuedByKubeAPIServerClientSigner(csr)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			newCSR := e.ObjectNew.(*certificatesv1.CertificateSigningRequest)
+			// agent certificate is signed by "kubernetes.io/kube-apiserver-client" signer
+			return isIssuedByKubeAPIServerClientSigner(newCSR)
+		},
+		DeleteFunc:  func(event.DeleteEvent) bool { return false },
+		GenericFunc: func(event.GenericEvent) bool { return false },
+	}
+
+	return controllerruntime.NewControllerManagedBy(mgr).
+		Named(csrApprovingController).
+		For(&certificatesv1.CertificateSigningRequest{}, builder.WithPredicates(predicateFunc)).
+		Complete(a)
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+// The Controller will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (a *AgentCSRApprovingController) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
+	klog.V(4).Infof("Reconciling for CertificateSigningRequest %s", req.Name)
+
+	// 1. get latest CertificateSigningRequest
+	var csr *certificatesv1.CertificateSigningRequest
+	var err error
+	if csr, err = a.client.CertificatesV1().CertificateSigningRequests().Get(ctx, req.Name, metav1.GetOptions{}); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Infof("no need to reconcile CertificateSigningRequest for it not found")
```

Comment:
```suggestion
			klog.Infof("no need to reconcile CertificateSigningRequest for it not found")
```
Missing object key in log.